src/management-system-v2/lib/data/legacy/folders.ts

```diff
@@ -0,0 +1,225 @@
+  if (!folderData) throw new Error('Folder not found');
+
+  if (ability && !ability.can('view', toCaslResource('Folder', folderData.folder)))
+    throw new Error('Permission denied');
+
+  return folderData.folder;
+}
+
+export function getFolderChildren(folderId: string, ability?: Ability) {
+  const folderData = foldersMetaObject.folders[folderId];
+  if (!folderData) throw new Error('Folder not found');
+
+  if (ability && !ability.can('view', toCaslResource('Folder', folderData.folder)))
+    throw new Error('Permission denied');
+
+  return folderData.children;
+}
+
+export function createFolder(folderInput: z.infer<typeof FolderSchema>, ability?: Ability) {
+  const folder = FolderSchema.parse(folderInput);
+  if (!folder.id) folder.id = v4();
+
+  // Checks
+  if (ability && !ability.can('create', toCaslResource('Folder', folder)))
+    throw new Error('Permission denied');
+
+  if (foldersMetaObject.folders[folder.id]) throw new Error('Folder already exists');
+
+  const parentFolderData = folder.parentId && foldersMetaObject.folders[folder.parentId];
+  if (folder.parentId) {
+    if (!parentFolderData) throw new Error('Parent folder does not exist');
+
+    if (parentFolderData.folder.environmentId !== folder.environmentId)
+      throw new Error('Parent folder is in a different environment');
+  } else {
+    if (foldersMetaObject.rootFolders[folder.environmentId])
+      throw new Error(`Environment ${folder.environmentId} already has a root folder`);
+  }
+
+  // Store
+  const newFolder = {
+    ...folder,
+    createdAt: new Date().toISOString(),
+    updatedAt: new Date().toISOString(),
+  } as Folder;
+
+  foldersMetaObject.folders[folder.id] = { folder: newFolder, children: [] };
+
+  if (parentFolderData) parentFolderData.children.push(newFolder);
+  else foldersMetaObject.rootFolders[newFolder.environmentId] = newFolder;
+
+  store.add('folders', newFolder);
+
+  return newFolder;
+}
+
+/** Deletes a folder and every child recursively */
+export function deleteFolder(folderId: string, ability?: Ability) {
+  // NOTE: maybe the ability should do this recursive check
+  const folderData = foldersMetaObject.folders[folderId];
+  if (!folderData) throw new Error('Folder not found');
+
+  _deleteFolder(folderData, ability);
+}
+
+/** internal function to delete folders from bottom to top */
+function _deleteFolder(
+  folderData: NonNullable<(typeof foldersMetaObject)['folders'][string]>,
+  ability?: Ability,
+) {
+  if (ability && !ability.can('delete', toCaslResource('Folder', folderData.folder)))
+    throw new Error('Permission denied');
+
+  for (const child of folderData.children) {
+    const childData = foldersMetaObject.folders[child.id];
+    if (!childData) throw new Error('Inconsistency in folder structure: child folder not found');
+
+    _deleteFolder(childData, ability);
+  }
+
+  // TODO: remove processes
+
+  delete foldersMetaObject.folders[folderData.folder.id];
+  if (!folderData.folder.parentId)
```

Comment:
If the folder has a parent shouldn't it be removed from the childrens array of said parent?

Reply:
You're totally right, I forgot to do that, I will also cover this in the tests in the next pr